03_execution/run-folder.py

```diff
@@ -1,166 +1,211 @@
+    # The example files in this repository start with \xEF\xBB\xBF See UTF-8 BOM
+    # If read normally these become part of the first header name
+    # utf-8-sig reads this correctly and also handles the case when they don't exist
+    with open(csv_path, "r", encoding="utf-8-sig") as csv_file:
         # pass the file object to reader() to get the reader object
-        csv_reader = reader(read_obj)
-
-        # Iterate over each row in the csv using reader object
-        lineNumber = 0
-        for row in csv_reader:
-            # row variable is a list that represents a row in csv
-            # break up the list of rgb values 
-            # remove the first item
-            if lineNumber > 0:
-                parsed_row = []
-                row.pop(0)
-                chunked_list = list(chunks(row, 3))
-                for element_num in range(len(chunked_list)):
-                    # this is a single light 
-                    r = float(chunked_list[element_num][0])
-                    g = float(chunked_list[element_num][1])
-                    b = float(chunked_list[element_num][2])
-                    light_val = (g, r, b)  # these LED lights take GRB color for some reason!
-                    # turn that led on
-                    parsed_row.append(light_val)
-                
-                # append that line to lightArray 
-                lightArray.append(parsed_row)
-            
-            lineNumber += 1
-
-    #print("Finished Parsing file " + csvFile)
-    
-    return lightArray
-
-
-# Given a light sequence, it plays it on the tree
-def playLights(neop, lights):
-    f = 0
-    for frame in lights:
-        LED = 0
-        while LED < NUMBEROFLEDS:
-            neop[LED] = frame[LED]
-            LED += 1
-        neop.show()
-        f += 1
-
-
-
-
-# Init the neopixel
-pixels = neopixel.NeoPixel(board.D18, NUMBEROFLEDS, auto_write=False)
-
-# get foldername from CLI arguments
-folder_path = sys.argv[1]
-
-# load csv files
-csv_files = []
-for file in os.listdir(folder_path):
-    if file.endswith(".csv"):
-        csv_files.append(file)
-
-
-# Parse all the sequences at the beginning (its a heavy process for the pi)
-sequences = []
-for file in csv_files:
-    full_path = os.path.join(folder_path, file)
-    if os.path.isfile(full_path):
-        lights = getLights(full_path)
-        sequences.append(lights)
-
-# Play sequences in a loop
-while True:
-    id = 0
-    for lights in sequences:
-        prev = sequences[id - 1]
-        print("Playing file " + csv_files[id])
-        for i in range(0, LOOPS_PER_SEQUENCE):
-            lights_now = []
-            if i != (LOOPS_PER_SEQUENCE - 1):
-                print("Loop " + str(i + 1) + " from " + str(LOOPS_PER_SEQUENCE), end='\r')
-                if i == 0:
-                    lights_now = blend_lights(prev, lights, TRANSITION_FRAMES)
-                else:
-                    lights_now = lights
+        csv_reader = csv.reader(csv_file)
```

Comment:
I see a lot of duplicated code between the run.py and this run-folder.py. Could we not just create a library file, which we import in both files? So, when we find a bug, we only need to change 1 file.

Otherwise, there will be soon an issue with different versions of the two files.

Reply:
Yes it absolutely can. I was just designing this for the case where someone may just want to download and run a single file.

Reply:
There was already a large amount of duplicated code between these two files so I assumed they wanted it runnable as a single file

Reply:
@garciadelcastillo Is this true, or should we avoid duplicated code? And make it better..

Reply:
I have moved all of the duplicated code and some other utility code to `run_utils.py`
This includes the csv parsing code and the frame drawing code
https://github.com/GSD6338/XmasTree/pull/4/commits/3368a0551a29fbab5f24f7c19ae83a7bc7f58fde#diff-2225fc874edd373fea2965452dc470310944e6e05210c38d0d4b6f48f644e03e

Reply:
Great!